### PR TITLE
Enable resizing the panes with the keyboard.

### LINKED ISSUE #991: Panes should be resizable with the keyboard.
Press a keybinding to adjust the width/height of a pair of panes. 

This comes with extra work - we have to have each parent pane track the relative size of each child, and then when the window is resized, recursively resize each control to match the absolute size to the previous proportions. Currently, we're just using two row/cols of equal width. 

Start with some simple amount, like 5% at a time. Adjusting by a single char at a time would only be possible if both panes were the same font size, which isn't guaranteed.

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Adds the ability to resize panes with the keyboard. 

## PR Checklist
* [x] Closes #991
* [x] I work here
* [ ] Tests added/passed - this box makes me cry
* [x] I actually kinda just went for this one. 

## Detailed Description of the Pull Request / Additional comments

This is accomplished by making the Column/RowDefinitions for a Pane use `GridLengthHelper::FromPixels` to set their size. We store a pair of floats that represents the relative amount that each pane takes out of the parent pane. When the window is resized, we use that percentage to figure out the new size of each child in pixels, and manually size each column. 

Then, when the user presses the keybindings for resizePane{Left/Right/Up/Down}, we'll adjust those percentages, and resize the rows/cols as appropriate.

Currently, each pane adjusts the width/height by 5% of the total size at a time. I am not in love with this, but it works for now. I think when we get support for keybindings with arbitrary arg blobs, then we could do either a percent movement, or a number of characters at a time. The number of characters one would be trickier, because we'd have to get the focused control, and get the number of pixels per character, as adjacent panes might not have the same font sizes.

## Validation Steps Performed

Opened a bunch of panes and tried moving them with the keybindings

### CODE REVIEW

File: src/cascadia/TerminalApp/Pane.h [hunk 1]
```diff
@@ -75,11 +80,49 @@ class Pane : public std::enable_shared_from_this<Pane>
...
 
     void _FocusFirstChild();
     void _ControlClosedHandler();
+
+    std::pair<float, float> _GetPaneSizes(const float& fullSize);
+
+    // Function Description:
+    // - Returns true if the given direction can be used with the given split
+    //   type.
+    // - This is used fow pane resizing (which will need a pane separator
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.h]: >fow [](start = 22, length = 3)

typo: for

File: src/cascadia/TerminalApp/Pane.cpp [hunk 2]
```diff
@@ -36,6 +38,155 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+//   we'll be increasing the size of the first of our children. Else, we'll be
+//   decreasing the size of our first child.
+// Return Value:
+// - <none>
+void Pane::_DoResize(const Direction& direction)
+{
+    if (!DirectionMatchesSplit(direction, _splitState))
+    {
+        return;
+    }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp]: (This is gonna sound confusing. Bear with me.) So, as of now, we don't do anything if you try resizing not-the direction of the split? Example: if `SplitDirection::Horizontal` and `Direction::Left` --> do nothing?

Why not find the nearest valid separator and adjust the size of all its children? It's definitely more complex but doable and I feel like it makes sense.

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/Pane.cpp]: Or, what if it becomes more like "resize greater" and "resize lesser", and which one it impacts is based on how the current terminal's pane's split?

File: src/cascadia/TerminalApp/Pane.cpp [hunk 3]
```diff
@@ -36,6 +38,155 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+//   currently focused pane, that's also in the correct direction to be moved.
+//   If there isn't such a separator, then this method returns false, as we
+//   couldn't handle the resize.
+// Arguments:
+// - direction: The direction to move the separator in.
+// Return Value:
+// - true if we or a child handled this resize request.
+bool Pane::ResizePane(const Direction& direction)
+{
+    // If we're a leaf, do nothing. We can't possibly have a descentant with a
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp]: >descentant [](start = 61, length = 10)

typo: descendant

File: src/cascadia/TerminalApp/Pane.cpp [hunk 4]
```diff
@@ -36,6 +38,155 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+
+    // Check if either our first or second child is the currently focused leaf.
+    // If it is, and the requested resize direction matches our separator, then
+    // we're the pane that needs to adjust its separator.
+    // If our separator is the wrong direction, then we can't handle it.
+    const bool firstIsFocused = _firstChild->_IsLeaf() && _firstChild->_lastFocused;
+    const bool secondIsFocused = _secondChild->_IsLeaf() && _secondChild->_lastFocused;
+    if (firstIsFocused || secondIsFocused)
+    {
+        if (Pane::DirectionMatchesSplit(direction, _splitState))
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp]: >Pane:: [](start = 12, length = 6)

nit: do we need the `Pane::` prefix here?

File: src/cascadia/TerminalApp/Pane.cpp [hunk 5]
```diff
@@ -36,6 +38,155 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+        {
+            _DoResize(direction);
+            return true;
+        }
+        return false;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Pane.cpp]: It seems like we're checking `DirectionMatchesSplit()` twice (in `ResizePane` and `_DoResize()`). Why not just have `_DoResize()` return a bool? Then you can replace all of this with `return _DoResize()`

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: I think I like this solution the most. Plus we could take this entire block:
```c++
            const bool handled = _secondChild->ResizePane(direction);
            if (handled)
            {
                return true;
            }
            else if (Pane::DirectionMatchesSplit(direction, _splitState))
            {
                _DoResize(direction);
                return true;
            }
            return false;
```
Into just:
```c++
            return _secondChild->ResizePane(direction) || _DoResize(direction);
```

File: src/cascadia/TerminalApp/Pane.cpp (hunk 4 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: Apparently not

File: src/cascadia/TerminalApp/Pane.cpp (hunk 2 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/Pane.cpp]: Ignore my statement above... Started playing with it after hooking it up to some temp key-bindings and it does exactly what I'm thinking of haha.

**Comment by DHowett-MSFT** [comment]
If you have this layout:

```
+--------------------------------+-----------------------------------+
|                                |                                   |
| *PARENT*                       | *SIBLING*                         |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                +-----------------------------------+
|                                |                                   |
|                                | *ACTIVE*                          |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
|                                |                                   |
+--------------------------------+-----------------------------------+
```

and you resize `<-->`, does it bubble up to the parent pane? Even if the horizontal split is nested inside the vertical one?

**Comment by carlos-zamora** [comment]
> If you have this layout:
> 
> ```
> +--------------------------------+-----------------------------------+
> |                                |                                   |
> | *PARENT*                       | *SIBLING*                         |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                +-----------------------------------+
> |                                |                                   |
> |                                | *ACTIVE*                          |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> |                                |                                   |
> +--------------------------------+-----------------------------------+
> ```
> 
> and you resize `<-->`, does it bubble up to the parent pane? Even if the horizontal split is nested inside the vertical one?

Yup. Pretty neat :)

**Comment by zadjii-msft** [comment]
@DHowett-MSFT you bet it does :)

File: src/cascadia/TerminalApp/Pane.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: Lol yea np :P

File: src/cascadia/TerminalApp/Pane.cpp [hunk 6]
```diff
@@ -554,6 +748,9 @@ void Pane::_DoSplit(SplitState splitType, const GUID& profile, const TermControl
 
     _splitState = splitType;
 
+    _firstPercent = { .50f };
+    _secondPercent = { .50f };
+
```
Comment (dlong11) [inline src/cascadia/TerminalApp/Pane.cpp]: Constant for the .05f?

File: src/cascadia/TerminalApp/Pane.h [hunk 7]
```diff
@@ -72,11 +77,51 @@ class Pane : public std::enable_shared_from_this<Pane>
...
+    bool _DoResize(const winrt::TerminalApp::Direction& direction);
 
     void _CloseChild(const bool closeFirst);
 
     void _FocusFirstChild();
     void _ControlClosedHandler();
+
+    std::pair<float, float> _GetPaneSizes(const float& fullSize);
+
+    winrt::Windows::Foundation::Size _GetMinSize();
```
Comment (dlong11) [inline src/cascadia/TerminalApp/Pane.h]: Can this method be const?

File: src/cascadia/TerminalControl/TermControl.h [hunk 8]
```diff
@@ -43,6 +43,8 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         void CopySelectionToClipboard(bool trimTrailingWhitespace);
         void Close();
         bool ShouldCloseOnExit() const noexcept;
+        Windows::Foundation::Size CharacterDimensions();
+        Windows::Foundation::Size MinimumSize();
```
Comment (dlong11) [inline src/cascadia/TerminalControl/TermControl.h]: Can these methods be const?

**Comment by zadjii-msft** [comment]
@adiviness @DHowett-MSFT @miniksa Can I get a second on this at some point? I'll have another PR to follow this one for moving pane focus

File: src/cascadia/TerminalApp/Pane.cpp [hunk 9]
```diff
@@ -36,6 +38,154 @@ Pane::Pane(const GUID& profile, const TermControl& control, const bool lastFocus
...
+    _CreateRowColDefinitions(newSize);
+
+    if (_splitState == SplitState::Vertical)
+    {
+        const auto paneSizes = _GetPaneSizes(width);
+
+        Size firstSize{ paneSizes.first, height };
+        Size secondSize{ paneSizes.second, height };
+        _firstChild->ResizeContent(firstSize);
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: Eh, yea, I'm not too worried about that. I think the code's more annoying to read the other way.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 10]
```diff
@@ -1065,6 +1068,13 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         size.cx = static_cast<long>(newWidth);
         size.cy = static_cast<long>(newHeight);
 
+        // Don't actually resize so small that a single character wouldn't fit
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControl.cpp]: It actually doesn't throw here, it waits till the renderer tries to come through and paint the line, and throws then.

File: src/cascadia/TerminalApp/Pane.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Pane.cpp]: #wontfix

**Pull request merged**